src/main/java/net/openhft/chronicle/wire/scoped/ScopedThreadLocal.java

```diff
@@ -0,0 +1,157 @@
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.reflect.Array;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * A scoped thread-local resource.
+ * <p>
+ * This is used for small, tightly-scoped thread-local resource "pools", a safer alternative
+ * to a thread-local singleton.
+ * <p>
+ * Example of use:
+ * <pre>{@code
+ *   try (ScopedResource<Wire> sharedWire = sharedWireScopedThreadLocal.get()) {
+ *     Wire wire = sharedWire.get();
+ *     // ... do something with the wire
+ *   } // it is returned for use by inner scopes here
+ * }</pre>
+ *
+ * @param <T>
+ */
+public class ScopedThreadLocal<T> {
+
+    private final Supplier<T> supplier;
+    private final Consumer<T> onAcquire;
+    private final CleaningThreadLocal<SimpleStack> instancesTL;
+    private final boolean useWeakReferences;
+
+    /**
+     * Constructor
+     *
+     * @param supplier     The supplier of new instances
+     * @param maxInstances The maximum number of instances that will be retained for re-use
+     */
+    public ScopedThreadLocal(Supplier<T> supplier, int maxInstances) {
+        this(supplier, ScopedThreadLocal::noOp, maxInstances);
+    }
+
+    /**
+     * Constructor
+     *
+     * @param supplier     The supplier of new instances
+     * @param onAcquire    A function to run on each instance upon it's acquisition
+     * @param maxInstances The maximum number of instances that will be retained for re-use
+     */
+    public ScopedThreadLocal(@NotNull Supplier<T> supplier, @NotNull Consumer<T> onAcquire, int maxInstances) {
+        this(supplier, onAcquire, maxInstances, false);
+    }
+
+    /**
+     * Constructor
+     *
+     * @param supplier          The supplier of new instances
+     * @param onAcquire         A function to run on each instance upon it's acquisition
+     * @param maxInstances      The maximum number of instances that will be retained for re-use
+     * @param useWeakReferences Whether to allow resources to be garbage collected when they're not in use
+     */
+    public ScopedThreadLocal(@NotNull Supplier<T> supplier, @NotNull Consumer<T> onAcquire, int maxInstances, boolean useWeakReferences) {
+        this.supplier = supplier;
+        this.onAcquire = onAcquire;
+        this.instancesTL = CleaningThreadLocal.withCloseQuietly(() -> new SimpleStack(maxInstances));
+        this.useWeakReferences = useWeakReferences;
+    }
+
+    /**
+     * Get a scoped instance of the shared resource
+     *
+     * @return the {@link ScopedResource}, to be closed once it is finished being used
+     */
+    public ScopedResource<T> get() {
+        final SimpleStack scopedThreadLocalResources = instancesTL.get();
+        AbstractScopedResource<T> instance;
+        if (scopedThreadLocalResources.isEmpty()) {
+            instance = createNewResource();
+        } else {
+            instance = scopedThreadLocalResources.pop();
+        }
```

Comment:
Note here that the instance created/acquired is removed from the `SimpleStack`, so if the `get()`-ter doesn't return the instance (by calling `ScopedResource#close()`), it'll end up being GC'd. If this is approved I might add a `Cleaner` that confirms the resource was discarded correctly and warns otherwise.

Reply:
The alternative would be keeping a reference to it when it's checked out, but that could lead to resource leaks. I think the instance being GC'd rather than being returned to the pool is probably preferable.